db/compaction/compaction_picker_fifo.cc

```diff
@@ -125,16 +125,27 @@ Compaction* FIFOCompactionPicker::PickSizeCompaction(
     const std::string& cf_name, const MutableCFOptions& mutable_cf_options,
     const MutableDBOptions& mutable_db_options, VersionStorageInfo* vstorage,
     LogBuffer* log_buffer) {
-  const int kLevel0 = 0;
-  const std::vector<FileMetaData*>& level_files = vstorage->LevelFiles(kLevel0);
-  uint64_t total_size = GetTotalFilesSize(level_files);
+  // compute the total size and identify the last non-empty level
+  int last_level = 0;
+  uint64_t total_size = 0;
+  for (int level = 0; level < vstorage->num_levels(); ++level) {
+    auto level_size = GetTotalFilesSize(vstorage->LevelFiles(level));
+    total_size += level_size;
+    if (level_size > 0) {
+      last_level = level;
+    }
+  }
+  const std::vector<FileMetaData*>& last_level_files =
+      vstorage->LevelFiles(last_level);
 
   if (total_size <=
           mutable_cf_options.compaction_options_fifo.max_table_files_size ||
-      level_files.size() == 0) {
-    // total size not exceeded
+      last_level_files.size() == 0) {
```

Comment:
The key idea here is to keep the existing FIFO compaction logic but use last_level_file to replace the old level_files (level 0) so that FIFO compaction will remove the oldest files from the oldest level.

Reply:
I think we probably can remove this check. If last_level_files.size() == 0, then total_size should be 0. It was cheap to add the check level_files.size() == 0 but now the logic is more complicated so maybe we can remove it.